src/infra/dataset-update-flow-inmem/src/services/flow/flow_service_inmem.rs

```diff
@@ -0,0 +1,733 @@
+        trigger: FlowTrigger,
+    ) -> Result<FlowState, InternalError> {
+        let mut flow = Flow::load(flow_id, self.flow_event_store.as_ref())
+            .await
+            .int_err()?;
+        flow.add_trigger(self.time_source.now(), trigger)
+            .int_err()?;
+        flow.save(self.flow_event_store.as_ref()).await.int_err()?;
+        Ok(flow.into())
+    }
+
+    fn enqueue_flow(
+        &self,
+        flow_id: FlowID,
+        activation_time: DateTime<Utc>,
+    ) -> Result<(), InternalError> {
+        self.state
+            .lock()
+            .unwrap()
+            .time_wheel
+            .activate_at(activation_time, flow_id)?;
+        Ok(())
+    }
+
+    async fn schedule_flow_task(&self, flow: &mut Flow) -> Result<(), InternalError> {
+        let logical_plan = match &flow.flow_key {
+            FlowKey::Dataset(flow_key) => match flow_key.flow_type {
+                DatasetFlowType::Update => LogicalPlan::UpdateDataset(UpdateDataset {
+                    dataset_id: flow_key.dataset_id.clone(),
+                }),
+                DatasetFlowType::Compacting => unimplemented!(),
+            },
+            FlowKey::System(flow_key) => {
+                match flow_key.flow_type {
+                    // TODO: replace on correct logical plan
+                    SystemFlowType::GC => LogicalPlan::Probe(Probe {
+                        dataset_id: None,
+                        busy_time: Some(std::time::Duration::from_secs(20)),
+                        end_with_outcome: Some(TaskOutcome::Success),
+                    }),
+                }
+            }
+        };
+
+        let task = self
+            .task_scheduler
+            .create_task(logical_plan)
+            .await
+            .int_err()?;
+
+        flow.on_task_scheduled(self.time_source.now(), task.task_id)
+            .int_err()?;
+        flow.save(self.flow_event_store.as_ref()).await.int_err()?;
+
+        let mut state = self.state.lock().unwrap();
+        state
+            .pending_flows
+            .track_flow_task(flow.flow_id, task.task_id);
+
+        Ok(())
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl FlowService for FlowServiceInMemory {
+    /// Runs the update main loop
+    async fn run(&self) -> Result<(), InternalError> {
+        // Initial scheduling
+        self.initialize_auto_polling_flows_from_configurations()
+            .await?;
+
+        // Main scanning loop
+        loop {
+            // Do we have a timeslot scheduled?
+            let maybe_nearest_activation_time = {
+                let state = self.state.lock().unwrap();
+                state.time_wheel.nearest_activation_moment()
+            };
+
+            // Is it time to execute it yet?
+            if let Some(nearest_activation_time) = maybe_nearest_activation_time
+                && nearest_activation_time <= self.time_source.now()
```

Comment:
I had no idea you can unpack something with `if let` and then compare it in a single statement!

Reply:
This requires a feature flag:
`#![feature(let_chains)]`